### DISCUSSION TITLE
Just drop i386 already!

### DISCUSSION THREAD

**Discussion by TheAssassin** [AppImage/AppImageKit#1302] [General]
IA-32 is dead. Rightfully so. We're running out of build environments that still support it, most distributions like Ubuntu and Fedora dropped support for it years ago. CentOS 7 is not really maintained anymore and even if you believe in the 10-year LTS lifespan of a product whose successor release was dropped already, it's going to be end-of-life in just four months.

There is no reason to invest any time or effort into building binaries for this platform any longer. It didn't make a lot of sense a couple of years ago when I first proposed this, but by now, anybody should understand this.

I will begin to drop i386 builds from my own tools in December. For zsync2 and AppImageUpdate, given the fact that the previous build platform Ubuntu 18.04 has become EOL 8 months ago and the successor 20.04 (currently the oldest still-supported Ubuntu LTS release) does not provide such builds any more, I will also drop it there.

I recommend this to the entire community: don't waste time on IA-32. Those users should have changed to a 64-bit build already. And they know so. Some hardware is just not worth supporting. We need to make the best of our time. Maintaining dead technology is far from the best.

**Comment by TheAssassin** [comment]
For the record, for 32-bit ARMv7, the situation is drastically different. Almost all distributions still support it officially. We have hardware around that depends on this, and after all, it's much more recent than IA-32. The moment platforms like Raspberry Pi 1 are no longer supported by distributions (and, honestly, they are so much slower than all the 64-bit ARMv8 successors), we should be dropping these as well. At this point, building the binaries is easily possible with QEMU and Docker and usually doesn't require any additional effort (and the builds are just as fast as 64-bit ARM in QEMU).

**Comment by probonopd** [comment]
Let's look at this from two perspectives: Philosophically and pragmatically.

* _Philosophically_, I am strictly against dropping something that we already have working, _ever_. I am against "forced obsolescence" and I don't want to contribute to it. I have 2 1-st gen 32-bit Atom netbooks that still run perfectly fine, until people "drop i386". Heck, this whole forced software/hardware update sprial is what drove me away from the Mac toward open source in the first place. And let's not even begin to talk about "environmentally friendly".
* _Pragmatially_, I can fully understand that not everyone will want to build everything for 32-bit Intel anymore, just like not everyone has even started to do ARM builds (something that will probably become more relevant in the future). Personally, I like the Go language (golang) for how trivially easy it is to cross-compile for the various architectures, making this a complete non-issue there. Unfortunately it's not quite as easy with C/C++, although tools like the zig compiler might help there.

So what should we do?

My 2 cents:
* We should ensure that the AppImage runtime and "core" tools can work on all architectures, and provide builds for 32-bit and 64-bit Intel and ARM
* It is up to the maintainer(s) of tools in AppImageCommunity to decide which architectures to build for. AppImageCommunity is a "do-ocracy": The one who does decides (so, this would apply for AppImageUpdate as is in AppImageCommunity)

**Reply by TheAssassin** [reply]
> that we already have working

The point is, it stops working. I don't intend to build my own distribution, nor move to something like CentOS for no apparent reason which will go EOL in the next four months...

Yesterday, I found Debian still makes i386 builds, so maybe that's an option. But realistically, they'll stop doing so as well at some point. But that's the last resort then. If you provide us with a reasonable build platform, then migration might make sense depending on the required effort.

Edit: this is the project where I still build i386 CMake binaries for historic reasons: https://github.com/TheAssassin/prebuilt-cmake
I am going to drop distros there the minute they are no longer supported.

**Reply by probonopd** [reply]
> The point is, it stops working.

How would something "stop working" if one doesn't touch it? We can still build on the build environment that we are using now and can do so forever, can't we? I really don't like "moving targets".

**Reply by TheAssassin** [reply]
Docker images are removed. Distributions stop distributing binaries after a while. We don't back up entire distributions including the software. New binaries cannot be built any more.

**Reply by tkchia** [reply]
Hello all,

Thanks for the project.  In case it is any help, Ubuntu 24.04 for the x86-64 still includes a "cross"-compiling GCC, which can produce i686 (IA-32) programs &mdash; the command is `i686-linux-gnu-gcc`.  What is more, Linux/x86-64 can run the output IA-32 programs natively.

I expect though that producing programs which can run on an i386, and not just an i686, might be a challenge, since this will require a version of GNU libc (or another libc) which is restricted to only i386 instructions.

Thank you!

**Reply by probonopd** [reply]
i686 is fine for me. My main use case are 2008-ish Atom computers, e.g., netbooks.

**Reply by crueter** [reply]
> i686 is fine for me. My main use case are 2008-ish Atom computers, e.g., netbooks.

That you haven't touched in 15 years?

**Reply by RolandHughes** [reply]
> 
> The point is, it stops working. I don't intend to build my own distribution, nor move to something like CentOS for no apparent reason which will go EOL in the next four months...
> 
> Yesterday, I found Debian still makes i386 builds, so maybe that's an option. But realistically, they'll stop doing so as well at some point. But that's the last resort then. If you provide us with a reasonable build platform, then migration might make sense depending on the required effort.
> 

As someone who started out on real computers with real operating systems in the 1980s working as a midnight computer operator, I'm always mortified at people not performing bare metal backups. Of the six machines currently running in my office, 3 of them have manual bare metal backups performed using https://www.terabyteunlimited.com/image-for-linux/ and the other 3 have the Windows version of that software using automated backup and custom built recovery media. Functional Disaster Recovery Plans are in my DNA.

According to FOSS there are 15 Linux distros still supporting 32-bit.

https://itsfoss.com/32-bit-linux-distributions/

Not all go back to i386. Personally I use MX Linux as a primary dev system with Mate desktop. Having said that LMDE7 is pretty kick-butt. Sparky should be avoided. Last time I installed it on a netbook it was like watching molasess drip in the winter.

Twice now, over my 4+ decade career in IT, I have taken projects with entities under Publicis Groupe. They were working on IPOS (Intel Point of Sale). That "Explore This Computer" app you see running on machines. I worked on the Linux version. Lyle and Vanna sat in a room with everyone and explained, well, to the kids, not me, that the PC market is the used car market. Sure, they wanted to sell the $5000+ make a gamer drool desktops, but . . . the bulk of the non-data-center business was corporate desktop leases. 

Due to Microsoft's draconian licensing practices, most corporations in first world countries lease desktops on 3-5 year cycles. This makes the supplier responsible for all software licensing and hardware maintenance.  Once they come off-lease a small percentage finds their way onto eBay and other used computer markets. Gateway and Dell ingrained into consumers the idea that $300 will get them a "family" computer to do their budgets on and let the kids do there homework. It's more deeply entrenched that $5 foot longs. You can't get $5 foot longs (or even 6-inch) subs from Subway anymore, but you can still get $300 "family" computers, off-lease.

Dirty little secret. An i7-gen4 is massive overkill for a "family" computer. Honestly, Core-2 Duo is their "Happy Place."

The used computer market in first world countries isn't large enough to absorb all of the off-lease computers. They get packed into shipping containers and sent to Second World countries (now called Developing by the World Bank.) There are actual "big box" retailers in second world countries selling used computers. They have displays just like big box retailers and run the same "Explore this Computer" applications. It is customized for their sales and pricing. Think Poland, Ukraine, Latvia and former Soviet republics that American doesn't have technology embargos against.

As unbelievable as it may sound, the cycle repeats. Businesses in Second World countries lease computers due to software licensing and hardware support issues. A not insignificant percentage survive the lease. Guess what? They get shipped off to third world countries. In the big cities, where they have semi-reliable electricity they have retailers. Said retailers run the AMD or Microsoft "Explore this Computer" app because the last version of IPOS I worked on dropped 32-bit support. This is why AMD is achieving record sales and Intel is circling the financial drain.

https://www.tomshardware.com/pc-components/cpus/amds-desktop-pc-market-share-hits-a-new-high-as-server-gains-slow-down-intel-now-only-outsells-amd-2-1-down-from-9-1-a-few-years-ago

Unless you are willing to support the entire market chain you don't have a market. Leasing companies are switching to AMD desktops because AMD still supports the last mile. It's also illegal to bundle OS with hardware in many/most EU countries so Linux rules the retail market. That's also why you find so many on-line places selling legitimate digital download licenses for Windows and other Microsoft products at prices well under $50, many times under $20. (Yes, there are scams too). Second world countries won't pay retail price for MS.

No, I'm not making all this up. I'm old. I've seen a lot. Here's a current link to IndiaMart.

https://dir.indiamart.com/impcat/hp-desktop-computer/processor-core-2-duo-q15061312/

No shortage of Core 2 Duo machines there. Notice the hard drive sizes are mostly below 500GB. Guess what happened to all those tiny SSDs nobody buys in America anymore? Oh, you should not find any computers for sale in India with fax technology. I don't know why, but it is still illegal to sell/ship/give fax technology to anyone in India. Something to do with data compression.

Oh, you going to say the later Core 2 Duo were all 64-bit, fine. Hello Pentiums!

https://dir.indiamart.com/impcat/dell-desktop-computer/processor-intel-pentium-q15060932/

I don't know if anyone runs AppImages on it, but Intel introduced the Quark 32-bit in 2013 targeting IoT and embedded markets.

Where does the bulk of the world's e-waste end up? China? Nope! They stopped taking it.  India? Nope! It ends up in Poland

https://www.ecoport.pl/en/

I was shocked too. Doing research for the sequel to _[Twenty of Two The Infamous They - Five Nuclear Wars You Can Win](https://twentyoftwo.com/)_ is where I found that out. So the sequel, _Twenty of Two The Infamous They - Druids_ has a section named Polish Mother Drones. Made from recycled e-waste and working CPUs nobody wanted.

So, to summarize.

There are large numbers of people in third world countries still running 32-bit desktops. You will find many of the Linux desktop distros still supporting 32-bit have people from these countries contributing to the code base.

As Intel has learned, via the bankruptcy of the company, you have to support the last mile of consumption to survive.

Just my 0.002 cents

**Reply by crueter** [reply]
How heartwarming. They can use older versions instead of complaining then 😁

**Reply by RolandHughes** [reply]
> How heartwarming. They can use older versions instead of complaining then 😁

They don't. The get the new distros that still support 32-bit. They also get the new flatpaks, AppImages, etc. that support the 32-bit machine they currently have.

Thinking everyone must have the latest and greatest is why Intel is going out of business.

**Reply by crueter** [reply]
Oh, why didn't I think of that? Think of the people using 30 year old hardware!

**Reply by RolandHughes** [reply]
> Oh, why didn't I think of that? Think of the people using 30 year old hardware!

Ahhhh kids these days.

I work in the world of medical devices. Once we go through FDA 510K approval process, nothing can change in the development system. There is a drug manufacturer in California that has a PDP 11/70 running RSTS/E controlling a manufacturing line. The have lots of spare parts, and hold their breath. They are sole source so I pray I never get whatever disease it's for. Yes, I get a call because I've worked on RSTS/E on a PDP 11/70

Every couple of years I get a phone call from Harman (sp?) to go on-site to work on some heart device using OS/2 and Qt/3.

The sheer cost and effort of going through 510K means you never touch it.

As to the third world countries running 30 year old systems, they are providing the low-cost/free developers keeping the 32-bit support working in the distros previously listed. If they aren't coming here in droves it would be:

1. Your stuff won't run on their systems
2. They don't know about you.

Both situations can be fixed.

**Reply by probonopd** [reply]
I am still running computers from 1984. So no worries.

**Reply by crueter** [reply]
How sad. The medical device maintainers can do it themselves then.

**Reply by RolandHughes** [reply]
> I am still running computers from 1984. So no worries.

Wow! Do you have my NCR PC4 dual floppy computer? I built a wood enclosure for it so I could have an extra power supply and 20MEG hard drive!

First machine I ever bought. I see people in the retro markets bringing them back to life.

There are days I miss that ole thing. The ghosting green phosphorus monitor where you could hear the text display like tiny paint brushes on the glass. That and the chirping Seagate 20Meg made for calming sounds while debugging.
